server/cache_util/cache.py

```diff
@@ -17,14 +17,52 @@
 #  limitations under the License.
 ###############################################################################
 
-from cachetools import LRUCache, Cache, hashkey
+import six
+from cachetools import LRUCache, hashkey
+
 from .cachefactory import CacheFactory
 
 
 def strhash(*args, **kwargs):
     return str(hashkey(*args, **kwargs))
 
 
+def methodcache(key=None):
+    """
+    Decorator to wrap a function with a memoizing callable that saves results
+    in self.cache.  This is largely taken from cachetools, but uses a cache
+    from self.cache rather than a passed value.  If self.cache_lock is
+    present and not none, a lock is used.
+
+    :param key: if a function, use that for the key, otherwise use self.wrapKey.
+    """
+    def decorator(func):
+        @six.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            k = key(*args, **kwargs) if key else self.wrapKey(*args, **kwargs)
+            lock = getattr(self, 'cache_lock', None)
+            try:
+                if lock:
+                    with self.cache_lock:
+                        return self.cache[k]
+                else:
+                    return self.cache[k]
+            except KeyError:
+                pass  # key not found
+            v = func(self, *args, **kwargs)
+            try:
+                if lock:
+                    with self.cache_lock:
+                        self.cache[k] = v
+                else:
+                    self.cache[k] = v
+            except ValueError:
+                pass  # value too large
+            return v
+        return wrapper
+    return decorator
```

Comment:
Yes -- I copied the method from the `cachetools` module and then changed it for our use.  Since it is based on that code, I decided to leave the structure as it.